mmtk/src/api.rs

```diff
@@ -282,3 +284,49 @@ pub extern "C" fn get_finalized_object() -> ObjectReference {
         None => unsafe { Address::ZERO.to_object_reference() },
     }
 }
+
+thread_local! {
+    /// Cache all the pointers reported by the current thread.
+    static NMETHOD_SLOTS: RefCell<Vec<Address>> = RefCell::new(vec![]);
+}
+
+/// Report a list of pointers in nmethod to mmtk.
+#[no_mangle]
+pub extern "C" fn mmtk_add_nmethod_oop(addr: Address) {
+    NMETHOD_SLOTS.with(|x| x.borrow_mut().push(addr))
+}
+
+/// Register a nmethod.
+/// The c++ part of the binding should scan the nmethod and report all the pointers to mmtk first, before calling this function.
+/// This function will transfer all the locally cached pointers of this nmethod to the global storage.
+#[no_mangle]
+pub extern "C" fn mmtk_register_nmethod(nm: Address) {
+    let slots = NMETHOD_SLOTS.with(|x| {
+        if x.borrow().len() == 0 {
+            return None;
+        }
+        Some(x.replace(vec![]))
+    });
+    let slots = match slots {
+        Some(slots) => slots,
+        _ => return,
+    };
+    let mut roots = crate::CODE_CACHE_ROOTS.lock().unwrap();
+    crate::CODE_CACHE_ROOTS_SIZE.store(
```

Comment:
I'm unsure why you are using a load+store instead of a `fetch_add()` here (and `fetch_sub()` for unregister nmethod)

Reply:
Sure. But I thought `fetch_add(number, Ordering::Relaxed)` would be the same? Or at least as per my understanding of the documentation for `fetch_add()` [1]

\[1]: https://doc.rust-lang.org/std/sync/atomic/struct.AtomicUsize.html#method.fetch_add

Reply:
Yeah. `fetch_add` will result in a `lock add` instruction, and load-store will result in a simple `add`. 

However, for cases like this, the most intuitive way is to just use `fetch_add()`. If you are doing something that is different, it would be better to put some comments to it so people understand and won't attempt to 'fix/refactor' it.

Reply:
Sure I'll add a comment to this